Feat #34 : 게시물 목록(API) 작성

upqnu opened the pull request:
- 게시물 목록(API) 작성
- 게시물 목록의 test code 작성(MockMvc)
- dummy post 생성하는 service를 따로 분리하여 생성

upqnu committed:
Feat #34 : 게시물 목록(API) 작성
- 게시물 목록(API) 작성
- 게시물 목록의 test code 작성(MockMvc)
- dummy post 생성하는 service를 따로 분리하여 생성

U-jinLee commented on src/main/java/com/example/socialmediafeed/domain/post/controller/PostController.java:
서비스의 비즈니스 로직이 들어가는 부분은 Service 단에서 처리를 해주는 것이 맞습니다. 그 이유로는
1. Controller단과 Service단의 역할 분리
2. Transaction 처리와 관련 문제
그렇기 때문에 Controller 단에서 로직 처리는 지양해 주세요

U-jinLee commented on src/main/java/com/example/socialmediafeed/domain/post/controller/PostController.java:
해당 로직의 코드는 더 줄일 수 있을 것 같습니다.

U-jinLee commented on src/test/java/com/example/socialmediafeed/domain/post/controller/PostControllerTest.java:
아까 말씀드렸던 것처럼 Test의 공통부를 활용해서 짜는 코드를 만들어 주세요

upqnu replied:
Test 공통부 활용하는 방식으로 수정하겠습니다.

chickencoc reviewed (approved):
저는 보통 service 안에 기능 코드를 작성하고 controller는 간단하게 만드는 편인데, 혹시 PostController에 기능 코드를 작성하신 이유가 있으신가요??

U-jinLee reviewed (approved):
우선 기능 구현 부분에서 테스트 시 문제가 없다고 하시면 pr 하신 후에 차차 리팩토링 하는 방법도 좋을 것 같습니다.

upqnu committed:
Refactor #34 : 게시물 목록 - controller 단순화

- PostController.getPosts 메서드의 pagination 및 검색 관련 로직을 service로 이동하여 단순화시킴

upqnu committed:
Refactor #34 : 게시물 목록 - repository 수정

-repository 선언부에 QuerydslPredicateExecutor 인터페이스를 상속
- queryDSL 문법에 맞게 @Query 작성
- hashtag 매개변수에 따라 게시물을 검색하고, List<Post>로 검색 결과를 반환

upqnu committed:
Refactor #34 : 게시물 목록 - service 수정

- controller에 작성되어 있던 검색 및 pagination 로직을 옮겨서 작성
- EntityManager 주입받아 JPAQuery 객체를 생성하여 검색어에 따른 검색결과를 찾는 동적 쿼리를 생성하고 실행

upqnu committed:
Refactor #34 : 게시물 목록 - Test 코드 수정

- IntegrationTest를 상속한 PostControllerTest 생성
- 전체 게시물 갯수가 20개인지 확인하는 testTotalPostsCount 작성(application run 직후 dummy 게시물 20개가 생성됨)
- 해쉬태그가 존재하는 'java'로 검색하면 200 OK를, 존재하지 않는 'soccer'로 검색하면 204 (콘텐츠 없음)를 확인하는 testGetPosts 작성

upqnu committed:
Refactor #34 : 게시물 목록 - repository 수정

- LIKE 절에서 사용할 이스케이프 문자가 '!'로 잘못 설정 -> 의도했던 공백(' ')으로 수정

upqnu committed:
Refactor #34 : 게시물 목록 - controller 수정

- post가 존재하지 않을 때 NO_CONTENT 상태를 반환하는 if문 삭제

upqnu committed:
Refactor #34 : 게시물 목록 - service 수정

- 메서드 앞에 @Transactional 선언

upqnu committed:
Refactor #34 : 게시물 목록 - PostControllerTest 수정

- Test 클래스에 선언된 불필요한 4개의 annotation 삭제(상속받는 IntegrationTest에 이미 선언되어 있음)

upqnu commented:
게시물 목록에서 아래와 같이 수정 commit을 올렸습니다.
1. controller 메서드 내부 post가 존재하지 않을 때 NO_CONTENT 상태를 반환하는 if문 삭제
2. service에서 메서드 앞 @Transactional을 붙여줌
3. test클래스에 선언된 불필요한 annotation 제거
